File: genes/map/manifestations/ts/map.ts
/**
 * TypeScript manifestation of the map gene
 * Must extract to canonical λ-IR and match soul
 */

export function map<A, B>(xs: A[], f: (a: A) => B): B[] {
  const result = new Array<B>(xs.length);
  for (let i = 0; i < xs.length; i++) {
    result[i] = f(xs[i]);
  }
  return result;
}

// Metadata for verification
export const __gene = {
  name: 'map',
  lang: 'ts',
  extractsTo: 'λxs.λf.case(xs, nil:nil, cons(h,t):cons(f(h), map(t,f)))',
  soul: 'λ7a3f9b2c'
};

// Property tests
export const __laws = {
  lengthPreservation: <A, B>(xs: A[], f: (a: A) => B) => {
    return map(xs, f).length === xs.length;
  },
  
  identity: <A>(xs: A[]) => {
    const id = (x: A) => x;
    return JSON.stringify(map(xs, id)) === JSON.stringify(xs);
  },
  
  compositionFusion: <A, B, C>(xs: A[], f: (a: A) => B, g: (b: B) => C) => {
    const composed = (x: A) => g(f(x));
    const doubleMap = map(map(xs, f), g);
    const singleMap = map(xs, composed);
    return JSON.stringify(doubleMap) === JSON.stringify(singleMap);
  }
};

File: genes/map/manifestations/py/map.py
"""
Python manifestation of the map gene
Must extract to canonical λ-IR and match soul
"""

def map_gene(xs, f):
    """
    Pure functional map implementation
    """
    return [f(x) for x in xs]


# Metadata for verification
__gene__ = {
    'name': 'map',
    'lang': 'py',
    'extracts_to': 'λxs.λf.case(xs, nil:nil, cons(h,t):cons(f(h), map(t,f)))',
    'soul': 'λ7a3f9b2c'
}


# Property tests
class Laws:
    @staticmethod
    def length_preservation(xs, f):
        return len(map_gene(xs, f)) == len(xs)
    
    @staticmethod
    def identity(xs):
        id_fn = lambda x: x
        return map_gene(xs, id_fn) == xs
    
    @staticmethod
    def composition_fusion(xs, f, g):
        composed = lambda x: g(f(x))
        double_map = map_gene(map_gene(xs, f), g)
        single_map = map_gene(xs, composed)
        return double_map == single_map


# Export for cross-language testing
if __name__ == "__main__":
    # Test vectors
    test_xs = [1, 2, 3, 4, 5]
    test_f = lambda x: x * 2
    
    result = map_gene(test_xs, test_f)
    print(f"Input: {test_xs}")
    print(f"Result: {result}")
    print(f"Soul verified: {__gene__['soul']}")